docs/developer-guide/plugins.md

```diff
@@ -14,36 +14,47 @@ We recommend your custom rules adhere to our [rule conventions](rules.md) for:
 
 ## The anatomy of a plugin
 
+This example plugin disallows the word "foo" in selectors:
+
 ```js
-// Abbreviated example
-const stylelint = require("stylelint");
+import stylelint from "stylelint";
+
+const {
+  createPlugin,
+  utils: { report, ruleMessages, validateOptions }
+} = stylelint;
+
+const ruleName = "foo-org/selector-no-foo";
 
-const ruleName = "plugin/foo-bar";
-const messages = stylelint.utils.ruleMessages(ruleName, {
-  expected: "Expected ..."
+const messages = ruleMessages(ruleName, {
+  rejected: (selector) => `Unexpected "foo" within selector "${selector}"`
 });
+
 const meta = {
-  url: "https://github.com/foo-org/stylelint-foo/blob/main/src/rules/foo-bar/README.md"
-  // deprecated: true,
+  url: "https://github.com/foo-org/stylelint-selector-no-foo/blob/main/README.md"
 };
 
-const ruleFunction = (primaryOption, secondaryOptionObject) => {
-  return (postcssRoot, postcssResult) => {
-    const validOptions = stylelint.utils.validateOptions(
-      postcssResult,
-      ruleName,
-      {
-        /* .. */
-      }
-    );
+const ruleFunction = (primary, secondaryOptions) => {
+  return (root, result) => {
+    const validOptions = validateOptions(result, ruleName, {
+      actual: primary,
+      possible: [true]
+    });
 
-    if (!validOptions) {
-      return;
-    }
+    if (!validOptions) return;
+
+    root.walkRules((ruleNode) => {
+      const { selector } = ruleNode;
+
+      if (!selector.includes("foo")) return;
 
-    // ... some logic ...
-    stylelint.utils.report({
-      /* .. */
+      report({
+        result,
+        ruleName,
+        message: messages.rejected(selector),
+        node: ruleNode,
+        word: selector
```

Comment:
question

Should `index` and `endIndex` be mentioned somewhere?

Reply:
Good catch. I've tried adding a new description to the `report()` sections (f83157c92). Any thoughts?

Reply:
> Any thoughts?

> For example:

has no examples.

> _Do not use…

Maybe use an `[!IMPORTANT]` instead?

Reply:
Thanks. Fixed the typos and changed to `[!IMPORTANT]` (af6d74393)